nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/CoreInstance.java

```diff
@@ -329,4 +341,66 @@ public void run() {
         }.runUnrestricted();
     }
 
+    /**
+     * Performs a {@link Function} atomically, in a cluster-wide critical section.
+     *
+     * @param session an existing session
+     * @param obj the object on which the function should be applied
+     * @param keyFunction the function used to generate the key for locking/unlocking the object
+     * @param function the function to apply while the object is locked
+     * @return the result of the function
+     * @since 9.3
+     */
+    public static <T, R> R doAtomically(CoreSession session, T obj, BiFunction<CoreSession, T, String> keyFunction,
+            BiFunction<CoreSession, T, R> function) throws ConcurrentUpdateException, InterruptedException {
+
+        // Compute key
+        String key = keyFunction.apply(session, obj);
+        KeyValueStore kvStore = Framework.getService(KeyValueService.class).getKeyValueStore(DOCUMENT_LOCK);
+
+        try {
+            boolean locked = tryLock(key, kvStore);
+            if (!locked) {
+                throw new ConcurrentUpdateException(
+                        "Failed to acquire the lock on the object associated with the key " + key);
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt(); // restore interrupt
+            throw e;
```

Comment:
The caller has no interest in dealing with `InterruptedException` so just rethrow a `NuxeoException` here and remove the `throws InterruptedException` from the signature of `doAtomically`